### PR TITLE
Tensorflow Bring Your Own Container

### PR DESCRIPTION
**Pull request by ChoiByungWook**
Adapted the scikit learn notebook.

Vanilla TensorFlow with TensorFlow Serving.

### CODE REVIEW

File: advanced_functionality/tensorflow_bring_your_own/container/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+# Download tensorflow serving
+RUN echo "deb [arch=amd64] http://storage.googleapis.com/tensorflow-serving-apt stable tensorflow-model-server tensorflow-model-server-universal" | tee /etc/apt/sources.list.d/tensorflow-serving.list
+RUN curl https://storage.googleapis.com/tensorflow-serving-apt/tensorflow-serving.release.pub.gpg | apt-key add -
+RUN apt-get update && apt-get install tensorflow-model-server
+
+ENV PATH="/opt/ml/code:${PATH}"
+
+COPY /cifar10 /opt/ml/code
+WORKDIR /opt/ml/code
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/Dockerfile]: You can do it after merging: explain add a comment explaining /opt/ml/code

File: advanced_functionality/tensorflow_bring_your_own/container/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# https://docs.docker.com/compose/gettingstarted/#step-2-create-a-dockerfile
+FROM tensorflow/tensorflow:1.8.0-py3
+
+RUN apt-get update && apt-get install -y --no-install-recommends nginx curl
+
+# Download tensorflow serving
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/Dockerfile]: You can do it after merging: add link from tf serving docs https://www.tensorflow.org/serving/setup#installing_the_modelserver

File: advanced_functionality/tensorflow_bring_your_own/container/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# https://docs.docker.com/compose/gettingstarted/#step-2-create-a-dockerfile
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/Dockerfile]: add copyright

Comment (mvsusp) [reply advanced_functionality/tensorflow_bring_your_own/container/Dockerfile]: You can do it after this merging:  Explain better the link above or remove it.

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import functools
+
+import os
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py]: organize the imports

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+_MOMENTUM = 0.9
+
+# We use a weight decay of 0.0002, which performs better than the 0.0001 that
+# was originally suggested.
+_WEIGHT_DECAY = 2e-4
+
+_BATCHES_PER_EPOCH = NUM_EXAMPLES_PER_EPOCH_FOR_TRAIN / BATCH_SIZE
+
+
+def model_fn(features, labels, mode):
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py]: You can do it after merging: add docstring linking to the model_fn docs: https://www.tensorflow.org/guide/custom_estimators#write_a_model_function

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        train_op = None
+
+    return tf.estimator.EstimatorSpec(
+        mode=mode,
+        predictions=predictions,
+        loss=loss,
+        train_op=train_op)
+
+
+def serving_input_fn():
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py]: You can do it after merging: add link on how to create a tf serving model and explain a little bit https://www.tensorflow.org/guide/saved_model#build_and_load_a_savedmodel

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+def main(model_dir, data_dir, train_steps):
+    tf.logging.set_verbosity(tf.logging.INFO)
+    train(model_dir, data_dir, train_steps)
+
+
+if __name__ == '__main__':
+    args_parser = argparse.ArgumentParser()
+    args_parser.add_argument(
+        '--data-dir',
+        default='/opt/ml/input/data/training',
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py]: You can do later: add comment explaining the channel

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        type=str,
+        # required=True,
+        help='The directory where the CIFAR-10 input data is stored.')
+    args_parser.add_argument(
+        '--model-dir',
+        default='/opt/ml/model',
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py]: You can do later: explain opt ml model folder

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/nginx.conf [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+http {
+  server {
+    # configures the server to listen to the port 8080
+    # https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-inference-code.html#your-algorithms-inference-code-container-response
+    listen 8080 deferred;
+
+    # redirects requests from SageMaker to TF Serving
+    location /invocations {
+      proxy_pass http://localhost:8501/v1/models/cifar10_model:predict;
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/nginx.conf]: You can do later: explain how this line works.

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/resnet_model.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/resnet_model.py]: Fix the copyright

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve [hunk 11]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env python
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve]: copyright

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+# This file implements the hosting solution, which just starts TensorFlow Model Serving.
+
+from __future__ import print_function
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve]: remove unnecessary import.

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+def start_server():
+    print('Starting TensorFlow Serving.')
+
+    # link the log streams to stdout/err so they will be logged to the container logs
+    subprocess.check_call(['ln', '-sf', '/dev/stdout', '/var/log/nginx/access.log'])
+    subprocess.check_call(['ln', '-sf', '/dev/stderr', '/var/log/nginx/error.log'])
+
+    nginx = subprocess.Popen(['nginx', '-c', '/opt/ml/code/nginx.conf'])
+    tf_model_server = subprocess.call(['tensorflow_model_server',
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve]: You can do later: add comments explaining each subprocess call.

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/train [hunk 14]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/train]: Copyright

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/train [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+channel_name = 'training'
+training_path = os.path.join(input_path, channel_name)
+
+# default params
+training_script = 'cifar10.py'
+default_params = ['--model-dir', str(model_path)]
+
+
+# Execute your training algorithm.
+def _run(cmd):
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/train]: You can do later: docstring explaining this function.

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/train [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=os.environ)
+    stdout, stderr = process.communicate()
+
+    return_code = process.poll()
+    if return_code:
+        error_msg = 'Return Code: {}, CMD: {}, Err: {}'.format(return_code, cmd, stderr)
+        raise Exception(error_msg)
+
+
+def _hyperparameters_to_cmd_args(hyperparameters):
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/train]: You can do later: docstring explaining this function

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/train [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        cmd_args_list.append('--{}'.format(key))
+        cmd_args_list.append(value)
+
+    return cmd_args_list
+
+
+if __name__ == '__main__':
+    try:
+        with open(param_path, 'r') as tc:
+            training_params = json.load(tc)
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/container/cifar10/train]: you can do later: comment explaining what you are doing here.

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 18]
```diff
@@ -0,0 +1,814 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Building your own algorithm container\n",
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: Should you rename to TensorFlow Bring your own container?

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 19]
```diff
@@ -0,0 +1,814 @@
...
+    "##### The input\n",
+    "\n",
+    "* `/opt/ml/input/config` contains information to control how your program runs. `hyperparameters.json` is a JSON-formatted dictionary of hyperparameter names to values. These values will always be strings, so you may need to convert them. `resourceConfig.json` is a JSON-formatted file that describes the network layout used for distributed training. Since scikit-learn doesn't support distributed training, we'll ignore it here.\n",
+    "* `/opt/ml/input/data/<channel_name>/` (for File mode) contains the input data for that channel. The channels are created based on the call to CreateTrainingJob but it's generally important that channels match what the algorithm expects. The files for each channel will be copied from S3 to this directory, preserving the tree structure indicated by the S3 key structure. \n",
+    "* `/opt/ml/input/data/<channel_name>_<epoch_number>` (for Pipe mode) is the pipe for a given epoch. Epochs start at zero and go up by one each time you read them. There is no limit to the number of epochs that you can run, but you must close each pipe before reading the next epoch.\n",
+    "\n",
+    "##### The output\n",
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: You can do it later: explain /opt/ml/code

File: advanced_functionality/tensorflow_bring_your_own/utils/generate_cifar10_tfrecords.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/utils/generate_cifar10_tfrecords.py]: check copyright

File: advanced_functionality/tensorflow_bring_your_own/utils/generate_cifar10_tfrecords.py [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import os
+import sys
+
+import tarfile
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/utils/generate_cifar10_tfrecords.py]: organize imports

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 22]
```diff
@@ -0,0 +1,814 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### The parts of the sample container\n",
+    "\n",
+    "In the `container` directory are all the components you need to package the sample algorithm for Amazon SageMager:\n",
+    "\n",
+    "    .\n",
+    "    ├── Dockerfile\n",
+    "    ├── build_and_push.sh\n",
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: Where is the build and push script?

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 23]
```diff
@@ -0,0 +1,814 @@
...
+       "   'classes': 4}]}"
+      ]
+     },
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import cv2\n",
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: Make sure to test the notebook in SageMaker Notebooks. Is it possible that they do not have all modules installed.

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 24]
```diff
@@ -0,0 +1,814 @@
...
+    "import numpy\n",
+    "\n",
+    "from sagemaker.predictor import json_serializer, json_deserializer\n",
+    "\n",
+    "image = cv2.imread(\"/path/to/image.png\", 1)\n",
+    "\n",
+    "# resize, as our model is expecting images in 32x32.\n",
+    "image = cv2.resize(image, (32, 32))\n",
+    "\n",
+    "data = {'instances': numpy.asarray(image).astype(float).tolist()}\n",
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: You can do later. Explain the format of the payload, you can link tf serving docs here.

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 25]
```diff
@@ -0,0 +1,814 @@
...
+    "\n",
+    "predictor.accept = 'application/json'\n",
+    "predictor.content_type = 'application/json'\n",
+    "\n",
+    "predictor.serializer = json_serializer\n",
+    "predictor.deserializer = json_deserializer\n",
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: You can do later: explain the serializers and the accept and content type

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 26]
```diff
@@ -0,0 +1,814 @@
...
+    "\n",
+    "predictor.predict(data)"
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: You can do later: link to the predictor documentation in the python SDK

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 27]
```diff
@@ -0,0 +1,814 @@
...
+       "   'classes': 3}]}"
+      ]
+     },
+     "execution_count": 33,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "image = cv2.imread(\"/path/to/image.png\", 1)\n",
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: You can do later: add the image in the notebook.

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 28]
```diff
@@ -0,0 +1,814 @@
...
+    "predictor.deserializer = json_deserializer\n",
+    "\n",
+    "predictor.predict(data)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Making predictions using Boto3\n",
```
Comment (mvsusp) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: You can do it later: link to boto3 documentation.

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 29]
```diff
@@ -0,0 +1,814 @@
...
+    "- [How Amazon SageMaker interacts with your Docker container for training](https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo.html)\n",
+    "- [How Amazon SageMaker interacts with your Docker container for inference](https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-inference-code.html)\n",
+    "- [CIFAR-10 Dataset](https://www.cs.toronto.edu/~kriz/cifar.html)\n",
+    "- [SageMaker Python SDK](https://github.com/aws/sagemaker-python-sdk)\n",
+    "- [Dockerfile](https://docs.docker.com/engine/reference/builder/)\n",
+    "- [scikit-bring-your-own](https://github.com/awslabs/amazon-sagemaker-examples/blob/master/advanced_functionality/scikit_bring_your_own/scikit_bring_your_own.ipynb)"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (laurenyu) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: don't forget to remove the empty cell

File: advanced_functionality/tensorflow_bring_your_own/container/Dockerfile (hunk 1 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/Dockerfile]: Added

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py (hunk 7 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py]: Added

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py (hunk 8 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/cifar10/cifar10.py]: Added

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/nginx.conf (hunk 9 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/cifar10/nginx.conf]: added

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve (hunk 11 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve]: done.

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve (hunk 12 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve]: done.

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve (hunk 13 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/cifar10/serve]: done

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/train (hunk 14 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/cifar10/train]: done

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/train (hunk 16 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/cifar10/train]: done

File: advanced_functionality/tensorflow_bring_your_own/container/cifar10/train (hunk 17 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/container/cifar10/train]: done

File: advanced_functionality/tensorflow_bring_your_own/utils/generate_cifar10_tfrecords.py (hunk 21 above)
Comment (ChoiByungWook) [reply advanced_functionality/tensorflow_bring_your_own/utils/generate_cifar10_tfrecords.py]: done

File: advanced_functionality/tensorflow_bring_your_own/container/Dockerfile (hunk 3 above)
Comment (mvsusp) [reply advanced_functionality/tensorflow_bring_your_own/container/Dockerfile]: Still missing copyright

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 30]
```diff
@@ -0,0 +1,743 @@
...
+    "\n",
+    "_or_ I'm impatient, just [let me see the code](#The-Dockerfile)!\n",
+    "\n",
+    "## When should I build my own algorithm container?\n",
+    "\n",
+    "You may not need to create a container to bring your own code to Amazon SageMaker. When you are using a framework such as Apache MXNet or TensorFlow that has direct support in SageMaker, you can simply supply the Python code that implements your algorithm using the SDK entry points for that framework. This set of supported frameworks is regularly added to, so you should check the current list to determine whether your algorithm is written in one of these common machine learning environments.\n",
+    "\n",
+    "Even if there is direct SDK support for your environment or framework, you may find it more effective to build your own container. If the code that implements your algorithm is quite complex or you need special additions to the framework, building your own container may be the right choice.\n",
+    "\n",
+    "This walkthrough shows that it is quite straightforward to build your own container. So don't worry if there isn't direct SDK support for your environment.\n",
```
Comment (djarpin) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: I'd change to: "So you can still use SageMaker even if your use case is not covered by the deep learning containers that we've built for you."

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 31]
```diff
@@ -0,0 +1,743 @@
...
+    "\n",
+    "## Permissions\n",
+    "\n",
+    "Running this notebook requires permissions in addition to the normal `SageMakerFullAccess` permissions. This is because it creates new repositories in Amazon ECR. The easiest way to add these permissions is simply to add the managed policy `AmazonEC2ContainerRegistryFullAccess` to the role that you used to start your notebook instance. There's no need to restart your notebook instance when you do this, the new permissions will be available immediately.\n",
+    "\n",
+    "## The example\n",
+    "\n",
+    "In this example we show how to package a custom TensorFlow container with a Python example which works with the CIFAR-10 dataset and uses TensorFlow Serving for inference. This example is kept simple as the main point is to show surrounding structure that you must add to your own code to train and host it in Amazon SageMaker.\n",
+    "\n",
+    "The approach demonstrated here works for any language or environment. You need to choose the right tools for your environment to serve HTTP requests for inference, but good HTTP environments are available in every language these days.\n",
```
Comment (djarpin) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: Maybe provide some examples.

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 32]
```diff
@@ -0,0 +1,743 @@
...
+    "    │       └── <channel_name>\n",
+    "    │           └── <input data>\n",
+    "    ├── model\n",
+    "    │   └── <model files>\n",
+    "    └── output\n",
+    "        └── failure\n",
+    "\n",
+    "##### The input\n",
+    "\n",
+    "* `/opt/ml/input/config` contains information to control how your program runs. `hyperparameters.json` is a JSON-formatted dictionary of hyperparameter names to values. These values are always strings, so you may need to convert them. `resourceConfig.json` is a JSON-formatted file that describes the network layout used for distributed training. Since scikit-learn doesn't support distributed training, we ignore it here.\n",
```
Comment (djarpin) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: scikit mention probably isn't relevant here.

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 33]
```diff
@@ -0,0 +1,743 @@
...
+     "text": [
+      "Collecting opencv-python\n",
+      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/53/e0/21c8964fa8ef50842ebefaa7346a3cf0e37b56c8ecd97ed6bd2dbe577705/opencv_python-3.4.2.17-cp36-cp36m-manylinux1_x86_64.whl (25.0MB)\n",
+      "\u001b[K    100% |████████████████████████████████| 25.0MB 2.1MB/s eta 0:00:01\n",
+      "\u001b[?25hRequirement already satisfied: numpy>=1.11.3 in /home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages (from opencv-python) (1.14.5)\n",
+      "\u001b[31mdistributed 1.21.8 requires msgpack, which is not installed.\u001b[0m\n",
+      "Installing collected packages: opencv-python\n",
+      "Successfully installed opencv-python-3.4.2.17\n",
+      "\u001b[33mYou are using pip version 10.0.1, however version 18.0 is available.\n",
+      "You should consider upgrading via the 'pip install --upgrade pip' command.\u001b[0m\n"
```
Comment (djarpin) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: Is there a reason you didn't clear output here?

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 34]
```diff
@@ -0,0 +1,743 @@
...
+     "text": [
+      "INFO:sagemaker:Deleting endpoint with name: tensorflow_cifar10_example-2018-08-03-18-06-55-168\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Gracefully stopping... (press Ctrl+C again to force)\n",
+      "Stopping tmp3n0u5hj2_algo-1-HCRIC_1 ... \r\n",
```
Comment (djarpin) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: I'd clear output here as well.

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 35]
```diff
@@ -0,0 +1,743 @@
...
+    "sess = sage.Session()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Upload the data for training\n",
+    "\n",
+    "When training models with huge amounts of data, you typically need big data tools, like Amazon Athena, AWS Glue, or Amazon EMR, to create your data in S3.\n",
```
Comment (djarpin) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: Not sure I understand how this sentence ties in.

File: advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb [hunk 36]
```diff
@@ -0,0 +1,743 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "b'{\\n    \"predictions\": [\\n        {\\n            \"classes\": 3,\\n            \"probabilities\": [0.122724, 0.0958609, 0.0519071, 0.272535, 0.097384, 0.0535893, 0.0905842, 0.0250508, 0.123435, 0.0669298]\\n        }\\n    ]\\n}'\n"
```
Comment (djarpin) [inline advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb]: I think we should add some markdown describing how to interpret the predictions.

**Pull request merged**